amazon/ionbenchmark/ion_benchmark_cli.py

```diff
@@ -84,52 +100,169 @@
 write_memory_usage_peak = 0
 read_memory_usage_peak = 0
 
+JSON_PRIMARY_BASELINE = Format.JSON
+CBOR_PRIMARY_BASELINE = Format.CBOR2
+
+output_file = 'dump_output'
 
-# Generates benchmark code for simpleion load API
-def generate_simpleion_load_test_code(file, memory_profiling, iterator=False, single_value=False,
+
+# Generates benchmark code for simpleion load/loads APIs
+def generate_simpleion_read_test_code(file, memory_profiling, io_type, iterator=False, single_value=False,
                                       emit_bare_values=False):
-    if not memory_profiling:
-        if not iterator:
-            def test_func():
-                with open(file, "br") as fp:
-                    data = ion.load(fp, single_value=single_value, emit_bare_values=emit_bare_values)
-                return data
+    if io_type == Io_type.BUFFER.value:
+        with open(file, "br") as fp:
+            benchmark_data = fp.read()
+        if not memory_profiling:
+            if not iterator:
+                def test_func():
+                    data = ion.loads(benchmark_data, single_value=single_value, emit_bare_values=emit_bare_values)
+                    return data
+            else:
+                def test_func():
+                    it = ion.loads(benchmark_data, single_value=single_value, emit_bare_values=emit_bare_values,
+                                   parse_eagerly=False)
+                    while True:
+                        try:
+                            next(it)
+                        except StopIteration:
+                            break
+                    return it
         else:
-            def test_func():
-                with open(file, "br") as fp:
-                    it = ion.load(fp, single_value=single_value, emit_bare_values=emit_bare_values,
-                                  parse_eagerly=False)
+            if not iterator:
+                def test_func():
+                    tracemalloc.start()
+                    data = ion.loads(benchmark_data, single_value=single_value, emit_bare_values=emit_bare_values)
+                    global read_memory_usage_peak
+                    read_memory_usage_peak = tracemalloc.get_traced_memory()[1] / BYTES_TO_MB
+                    tracemalloc.stop()
+                    return data
```

Comment:
`test_func()` is what is being timed, correct? That means the cost of using `tracemalloc` is included in the results. Is there a way for us to extract this logic outside of the timed block?

Reply:
Yes, we execute it twice; once with `memory_profiling` enabled, and once with it disabled. 

We profile the memory before all performance benchmarking [here](https://github.com/amazon-ion/ion-python/blob/master/amazon/ionbenchmark/ion_benchmark_cli.py#L284-L287). Then we [benchmark](https://github.com/amazon-ion/ion-python/blob/master/amazon/ionbenchmark/ion_benchmark_cli.py#L289-L293) the performance separately.